File: Java/HeapSort.java
package Java;

public class HeapSort extends Sort{
	
	public static void main(String[] args){
		int[] data = buildDataSet(5);
		
		timeSort(new HeapSort(), data);
	}
	
	public HeapSort(){ super("Heap"); }
	
	private void deleteMin(int[] data, int currentSize){    
	    int min = data[0];
	    data[0] = data[currentSize--];
	    percolateDown(data, 0, currentSize);
	    data[currentSize + 1] = min;
	}

	private void heapify(int[] data, int currentSize){
	    for (int i = currentSize/2; i >= 0; i--)
	        percolateDown(data, i, currentSize);
	}

	private void percolateDown(int[] data, int node, int currentSize){
	    int temp = data[node];

	    while ( (node + 1) * 2 - 1 <= currentSize )
	    {
	        int child = (node + 1) * 2 - 1; //left child
	        
	        if ( child != currentSize && data[child + 1] > data[child] )
	            child++;
	        if ( data[child] > temp )
	            data[node] = data[child];
	        else
	            break;

	        node = child;
	    }
	    data[node] = temp;
	}
	
	@Override
	public void sort(int[] data){
	    int currentSize = data.length - 1;
	    
	    heapify(data, currentSize);

	    while ( currentSize > 0 )
	        deleteMin(data, currentSize--);
	}
}

File: javascript/heap-sort.js
function percolateDown(array, node, currentSize){
    let temp = array[node];

    while ( (node + 1) * 2 - 1 <= currentSize )
    {
        let child = (node + 1) * 2 - 1; //left child
        
        if ( child != currentSize && array[child + 1] > array[child] )
            child++;
        if ( array[child] > temp )
            array[node] = array[child];
        else
            break;

        node = child;
    }
    array[node] = temp;
}

function deleteMin(array, currentSize){    
    let min = array[0];
    array[0] = array[currentSize--];
    percolateDown(array, 0, currentSize);
    array[currentSize + 1] = min;
}

function heapify(array, currentSize){
    for (let i = Math.floor(currentSize/2); i >= 0; i--)
        percolateDown(array, i, currentSize);
}

function sort(array){
    let currentSize = array.length - 1;
    
    heapify(array, currentSize);

    while ( currentSize > 0 )
        deleteMin(array, currentSize--);
}
